File: 2421-number-of-good-paths/2421-number-of-good-paths.py
class Solution:
    def numberOfGoodPaths(self, vals, edges):
        edges.sort(key=lambda x: max(vals[x[0]], vals[x[1]]))

        goodPaths = n = len(vals)
        par = [i for i in range(n)]
        rank = [1]*n

        def find(p):
            while par[p] != p:
                par[p]= par[par[p]]
                p = par[p]
            return p

        for a,b in edges:
            parent_a, parent_b = find(a), find(b)

            if vals[parent_a] == vals[parent_b]:
                goodPaths += rank[parent_a] * rank[parent_b]
                par[parent_a] = parent_b
                rank[parent_b] += rank[parent_a]
            elif vals[parent_a] > vals[parent_b]:
                par[parent_b] = parent_a
            else:
                par[parent_a] = parent_b

        return goodPaths

File: 2421-number-of-good-paths/2421-number-of-good-paths.cs
public class Solution
{
    public int NumberOfGoodPaths(int[] vals, int[][] edges)
    {
        var sortedEdges = edges.OrderBy(x => Math.Max(vals[x[0]], vals[x[1]]));
        var n = vals.Length;
        var goodPaths = n;
        var par = Enumerable.Range(0, n).ToArray();
        var rank = Enumerable.Repeat(1, n).ToArray();

        foreach (var edge in sortedEdges)
        {
            var parentA = FindParent(edge[0], par);
            var parentB = FindParent(edge[1], par);

            if (vals[parentA] == vals[parentB])
            {
                goodPaths += rank[parentA] * rank[parentB];
                par[parentA] = parentB;
                rank[parentB] += rank[parentA];
            }
            else if (vals[parentA] > vals[parentB])
            {
                par[parentB] = parentA;
            }
            else
            {
                par[parentA] = parentB;
            }
        }

        return goodPaths;
    }

    private static int FindParent(int p, IList<int> par)
    {
        while (par[p] != p)
        {
            par[p] = par[par[p]];
            p = par[p];
        }

        return p;
    }
}